### ISSUE TITLE
Build issue with eigen (MappedSparseMatrix)

### ISSUE THREAD

**Issue by meerfrau** [ecmwf/eckit#92]
### What happened?

MappedSparseMatrix was [removed](https://gitlab.com/libeigen/eigen/-/merge_requests/725) from eigen 3.4.x

Therefore any build against it fails:
```
./eckit/linalg/sparse/LinearAlgebraEigen.cc:29:22: error: 'MappedSparseMatrix' in namespace 'Eigen' does not name a template type; did you mean 'SparseMatrix'?
   29 | using spm_t = Eigen::MappedSparseMatrix<Scalar, Eigen::RowMajor, Index>;
      |                      ^~~~~~~~~~~~~~~~~~
      |                      SparseMatrix
./eckit/linalg/sparse/LinearAlgebraEigen.cc: In member function 'virtual void eckit::linalg::sparse::LinearAlgebraEigen::spmv(const eckit::linalg::SparseMatrix&, const eckit::linalg::Vector&, eckit::linalg::Vector&) const':
./eckit/linalg/sparse/LinearAlgebraEigen.cc:45:5: error: 'spm_t' was not declared in this scope
   45 |     spm_t Ai(A.rows(), A.cols(), A.nonZeros(), const_cast<Index*>(A.outer()), const_cast<Index*>(A.inner()), const_cast<Scalar*>(A.data()));
      |     ^~~~~
./eckit/linalg/sparse/LinearAlgebraEigen.cc:49:10: error: 'Ai' was not declared in this scope; did you mean 'yi'?
   49 |     yi = Ai * xi;
      |          ^~
      |          yi
./eckit/linalg/sparse/LinearAlgebraEigen.cc: In member function 'virtual void eckit::linalg::sparse::LinearAlgebraEigen::spmm(const eckit::linalg::SparseMatrix&, const eckit::linalg::Matrix&, eckit::linalg::Matrix&) const':
./eckit/linalg/sparse/LinearAlgebraEigen.cc:62:5: error: 'spm_t' was not declared in this scope
   62 |     spm_t Ai(A.rows(), A.cols(), A.nonZeros(), const_cast<Index*>(A.outer()), const_cast<Index*>(A.inner()), const_cast<Scalar*>(A.data()));
      |     ^~~~~
./eckit/linalg/sparse/LinearAlgebraEigen.cc:66:10: error: 'Ai' was not declared in this scope; did you mean 'Ci'?
   66 |     Ci = Ai * Bi;
      |          ^~
      |          Ci
```

When I however build against a locally cloned eigen (3.2.1 or patched 3.3.9) at eckit/maths/eigen3 the build works, but unluckily it <s>segfaults</s>(?) SIGABRTs when running:

`./eckit-info`
```
Exception: Bad operator: ELF (String) method 'contains' not implemented  (./eckit/value/Content.cc +124 contains)
terminate called after throwing an instance of 'eckit::BadOperator'
  what():  Bad operator: ELF (String) method 'contains' not implemented
```

### What are the steps to reproduce the bug?

```
  cmake \
    -DCMAKE_INSTALL_PREFIX=/usr \
    -DCMAKE_BUILD_TYPE=production \
..
make
```

### Version

git master

### Platform (OS and architecture)

ArchLinux

### Organisation

earth friends

**Comment by pmaciel** [comment]
Can you write how it segfaults when running? So I can reproduce.

I routinely build and test with eigen 3.4.0 (on both macOS/ARM64 and RHEL/amd64) and I don't find the same issue(s).

**Comment by meerfrau** [comment]
[Here is an strace](https://pastebin.com/y7Ee2q1p) of eckit-info

> I routinely build and test with eigen 3.4.0 (on both macOS/ARM64 and RHEL/amd64) and I don't find the same issue(s).
May you please link me an RPM of the eigen 3.4.0 which you use? 

I've invested quite an effort to get it compiling. I could be mad at myself ;)

**Comment by pmaciel** [comment]
@meerfrau you're right that the Eigen::MappedSparseMatrix has been deprecated and removed from Eigen's master branch. But, note that the master branch is currently ahead of the latest tag (we, in principle and as you're assuming, always tag the latest commit to master/main)

Note that eckit's interfacing to Eigen complies with the latest tag (3.4.0) -- inclusively, I actually was using 3.4.0 (without paying much attention) for integration with the Research Department which are always ahead of our production stack. 

Obviously, a notice of deprecation is very important and will be heeded. At the moment however, without a formal release, there isn't an urgency.

**Comment by meerfrau** [comment]
@pmaciel You maybe haven't noticed it, but I quickly changed to eigen 3.3.9 to compile successfully. I was mostly concerned about the abort signal directly after starting eckit:

```
Exception: Bad operator: ELF (String) method 'contains' not implemented  (./eckit/value/Content.cc +124 contains)
terminate called after throwing an instance of 'eckit::BadOperator'
  what():  Bad operator: ELF (String) method 'contains' not implemented
```

I've created a zipped tarball, I hope that works for you. May you please check whether you get the same error on running?
[eckit_bin.tar.zip](https://github.com/ecmwf/eckit/files/13454497/eckit_bin.zip)

**Comment by pmaciel** [comment]
Unfortunately I don't have a setup able to run your build as-is. I don't know you you mean by eigen "patched" 3.3.9, and one the "What are the steps to reproduce the bug" there's the cmake variable CMAKE_BUILD_TYPE=production which actually fails configuration in eigen 3.3.9 (they have something special going on there that I don't think is correct). 

The failure you mention is not really symptomatic of eigen but instead of something else (?). I suspect your build directory is out of sync with the sources, and try first to do a clean build disabling Eigen (passing -DENABLE_EIGEN=OFF).

Berfore proceeding, can you try the above please?

**Comment by odlomax** [comment]
In case it's relevant, I've been using `Eigen::Map<SparseMatrix>` successfully with eckit `SparseMatrix` objects. Appears to be supported by Eigen 3.4.x https://www.eigen.tuxfamily.org/dox/classEigen_1_1Map_3_01SparseMatrixType_01_4.html

**Comment by xiangyi-wang** [comment]
I encountered the same problem and 3.3.9 can succeed

**Comment by wdeconinck** [comment]
It seems this was sorted by @pmaciel  with commit d01435a75 and released in version  [1.28.1](https://github.com/ecmwf/eckit/releases/tag/1.28.1)
I am not sure if this is backwards compatible with Eigen 3.3.9, and now requires Eigen 3.4.0 as minimum requirement.
@pmaciel could you comment?

**Comment by pmaciel** [comment]
I've just coinfirmed that eckit [1.28.1|eckit https://github.com/ecmwf/eckit/releases/tag/1.28.1], together with Eigen [3.3.9|https://gitlab.com/libeigen/eigen/-/archive/3.3.9/eigen-3.3.9.tar.bz2] work correctly. To configure such build with ecbuild + eckit, this works:
$ /path/to/ecbuild .. -DEigen3_ROOT=/path/to/eigen-3.3.9 -DENABLE_EIGEN=ON
$ # build
$ ctest -R eckit_test_linalg_sparse_backend_eigen
1/1 Test #55: eckit_test_linalg_sparse_backend_eigen ...   Passed    0.02 sec

The option -DENABLE_EIGEN=ON forces the finding of Eigen. For this older version, it is important to keep your own version of Eigen rather than let the configuration pick the system's, which in my case (macOS with homebrew) is 3.4.0_1 (which also succeeds, but that's besides the point.)
